apps/src/templates/BaseDialog.jsx

```diff
@@ -38,7 +38,11 @@ var BaseDialog = React.createClass({
   },
 
   focusDialog: function () {
-    if (this.props.isOpen) {
+    // Don't steal focus if the active element is already a descendant of the
+    // dialog - prevents focus loss on updates of open BaseDialog components.
+    const descendantIsActive = document.activeElement && this.refs.dialog &&
+        this.refs.dialog.contains(document.activeElement);
+    if (this.props.isOpen && !descendantIsActive) {
       this.refs.dialog.focus();
```

Comment:
This is the actual fix for the reported bug.  The `BaseDialog` component steals focus every time it's rendered, to make sure the <kbd>Esc</kbd> key can close the dialog.  This produced an undesirable behavior when the dialog was open and contained form fields - the fields would lose focus on every re-render, as the dialog itself claimed it at the end of the render process.  Thanks @Bjvanminnen for putting me on the right track here.

The solution is to check whether a descendant of the dialog already has focus, and do nothing in that case.